validation/hooks_stdin.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/opencontainers/runtime-tools/specerror"
+	"github.com/opencontainers/runtime-tools/validation/util"
+	uuid "github.com/satori/go.uuid"
+)
+
+func stdinStateCheck(outputDir, hookName string, expectedState rspecs.State) error {
+	var state rspecs.State
+	data, err := ioutil.ReadFile(filepath.Join(outputDir, hookName))
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(data, &state)
+	if err != nil {
+		return err
+	}
+
+	if state.Version != expectedState.Version {
+		return fmt.Errorf("wrong OCI specification version %q in the stdin of %s hook, expected %q", state.Version, hookName, expectedState.Version)
+	}
+
+	if state.ID != expectedState.ID {
+		return fmt.Errorf("wrong container ID %q in the stdin of %s hook, expected %q", state.ID, hookName, expectedState.ID)
+	}
+
+	if state.Bundle != expectedState.Bundle {
+		return fmt.Errorf("wrong bundle directory %q in the stdin of %s hook, expected %q", state.Bundle, hookName, expectedState.Bundle)
+	}
+
+	if hookName != "poststop" && state.Pid != expectedState.Pid {
+		return fmt.Errorf("wrong container process ID %q in the stdin of %s hook, expected %q", state.Version, hookName, expectedState.Version)
+	}
+
+	if !reflect.DeepEqual(state.Annotations, expectedState.Annotations) {
+		return fmt.Errorf("wrong annotations \"%v\" in the stdin of %s hook, expected \"%v\"", state.Annotations, hookName, state.Annotations)
+	}
+	return nil
+}
+
+func main() {
+	t := tap.New()
+	t.Header(0)
+
+	bundleDir, err := util.PrepareBundle()
+	if err != nil {
+		util.Fatal(err)
+	}
+	containerID := uuid.NewV4().String()
+	defer os.RemoveAll(bundleDir)
+
+	var containerPid int
+
+	annotationKey := "org.opencontainers.runtime-tools"
+	annotationValue := "hook stdin test"
+	g := util.GetDefaultGenerator()
+	outputDir := filepath.Join(bundleDir, g.Spec().Root.Path)
+	g.AddAnnotation(annotationKey, annotationValue)
+	g.AddPreStartHook(rspecs.Hook{
+		Path: filepath.Join(bundleDir, g.Spec().Root.Path, "/bin/sh"),
+		Args: []string{
+			"sh", "-c", fmt.Sprintf("timeout -t 1 cat > %s", filepath.Join(outputDir, "prestart")),
+		},
+	})
+	g.AddPostStartHook(rspecs.Hook{
+		Path: filepath.Join(bundleDir, g.Spec().Root.Path, "/bin/sh"),
+		Args: []string{
+			"sh", "-c", fmt.Sprintf("timeout -t 1 cat > %s", filepath.Join(outputDir, "poststart")),
+		},
+	})
+	g.AddPostStopHook(rspecs.Hook{
+		Path: filepath.Join(bundleDir, g.Spec().Root.Path, "/bin/sh"),
+		Args: []string{
+			"sh", "-c", fmt.Sprintf("timeout -t 1 cat > %s", filepath.Join(outputDir, "poststop")),
+		},
+	})
+	g.SetProcessArgs([]string{"true"})
+	config := util.LifecycleConfig{
+		BundleDir: bundleDir,
+		Config:    g,
+		Actions:   util.LifecycleActionCreate | util.LifecycleActionStart | util.LifecycleActionDelete,
+		PreCreate: func(r *util.Runtime) error {
+			r.SetID(containerID)
+			return nil
+		},
+		PreDelete: func(r *util.Runtime) error {
+			state, err := r.State()
+			if err != nil {
+				return err
+			}
+			containerPid = state.Pid
```

Comment:
I think this would be more reliable if we dropped the `state` call and collected this from the container process itself (where you currently just use `true`).

Reply:
`echo $$` will output the id inside the container. It is different with the state one - `It is 'seen' by host.`

Reply:
> `echo $$` will output the id inside the container. It is different with the state one - `It is 'seen' by host.`

Ah right, you'd need the PID namespace too to translate, and that would break down for VM-based runtimes anyway.  So I'm on board with your current `state` call.